lib/multiplexer/proxyline.go

```diff
@@ -361,3 +374,160 @@ func MarshalTLVs(tlvs []TLV) ([]byte, error) {
+	}
+	if len(signingCert) == 0 {
+		return trace.BadParameter("missing signing certificate")
+	}
+
+	signatureTLVs := []TLV{
+		{
+			Type:  PP2Type(PP2TeleportSubtypeSigningCert),
+			Value: signingCert,
+		},
+		{
+			Type:  PP2Type(PP2TeleportSubtypeJWT),
+			Value: signature,
+		},
+	}
+	signatureBytes, err := MarshalTLVs(signatureTLVs)
+	if err != nil {
+		return err
+	}
+
+	// If there's already signature among TLVs, we replace it
+	for i := range p.TLVs {
+		if p.TLVs[i].Type == PP2TypeTeleport {
+			p.TLVs[i].Value = signatureBytes
+			return nil
+		}
+	}
+
+	// Otherwise we append it
+	p.TLVs = append(p.TLVs, TLV{Type: PP2TypeTeleport, Value: signatureBytes})
+
+	return nil
+}
+
+// isSigned returns true if proxy line's TLV contains signature.
+// Does not take into account if signature is valid or not, just the presence of it.
+func (p *ProxyLine) isSigned() bool {
+	token, proxyCert, _ := p.getSignatureAndSigningCert()
+	return len(token) > 0 || proxyCert != nil
+}
+
+// getSignatureAndSigningCert finds signature and signing certificate in TLVs if they are present
+func (p *ProxyLine) getSignatureAndSigningCert() (string, []byte, error) {
+	var proxyCert []byte
+	var token string
+	for _, tlv := range p.TLVs {
+		if tlv.Type == PP2TypeTeleport {
+			teleportSubTLVs, err := UnmarshalTLVs(tlv.Value)
+			if err != nil {
+				return "", nil, err
+			}
+
+			for _, subTLV := range teleportSubTLVs {
+				if subTLV.Type == PP2Type(PP2TeleportSubtypeSigningCert) {
+					proxyCert = subTLV.Value
+				}
+
+				if subTLV.Type == PP2Type(PP2TeleportSubtypeJWT) {
+					token = string(subTLV.Value)
+				}
+			}
+			break
+		}
+	}
+	return token, proxyCert, nil
+}
+
+// VerifySignature checks that signature contained in the proxy line is securely signed.
+func (p *ProxyLine) VerifySignature(hostCACerts [][]byte, clock clockwork.Clock) (bool, error) {
+	// If there's no TLVs it can't be verified
+	if len(p.TLVs) == 0 {
+		return false, nil
+	}
+
+	token, proxyCert, err := p.getSignatureAndSigningCert()
+	if err != nil {
+		return false, err
+	}
+	if len(token) == 0 || proxyCert == nil {
+		return false, nil
+	}
+
+	signingCert, err := tlsca.ParseCertificatePEM(proxyCert)
+	if err != nil {
+		return false, err
+	}
+
+	roots := x509.NewCertPool()
+	for _, cert := range hostCACerts {
+		ok := roots.AppendCertsFromPEM(cert)
+		if !ok {
+			return false, nil
+		}
+	}
+
+	// Make sure that transmitted proxy cert is signed by one of the provided HostCAs
+	chains, err := signingCert.Verify(x509.VerifyOptions{Roots: roots})
+	if err != nil && !errors.As(err, &x509.UnknownAuthorityError{}) {
+		return false, err
+	}
+	if chains == nil {
+		return false, nil
+	}
+
+	identity, err := tlsca.FromSubject(signingCert.Subject, signingCert.NotAfter)
+	if err != nil {
+		return false, err
+	}
+	clusterName := identity.TeleportCluster
+
+	foundRole := checkForSystemRole(identity, types.RoleProxy)
+	if !foundRole {
+		return false, nil
```

Comment:
Maybe it would be helpful to emit a log message or an error here ! I imagine there's a chance a bad actor would try a certificate that does not belong to a proxy, and some visibility of this would be great for an operator/security team.